linera-service/src/linera-exporter/runloops/block_processor.rs

```diff
@@ -0,0 +1,461 @@
+    node: ProcessedBlock,
+    sibling_branch: usize,
+}
+
+impl<T> BlockProcessor<T>
+where
+    T: Storage + Clone + Send + Sync + 'static,
+{
+    pub(super) fn new(
+        storage: BlockProcessorStorage<T>,
+        queue_rear: UnboundedSender<BlockId>,
+        queue_front: UnboundedReceiver<BlockId>,
+    ) -> Self {
+        Self {
+            storage,
+            queue_rear,
+            queue_front,
+        }
+    }
+
+    pub(super) async fn run_with_shutdown<F>(
+        &mut self,
+        signal: F,
+        persistence_period: u16,
+    ) -> Result<(), ExporterError>
+    where
+        F: IntoFuture<Output = ()>,
+    {
+        let future = signal.into_future();
+        let mut pinned = Box::pin(future);
+
+        let mut interval = interval(Duration::from_secs(persistence_period.into()));
+        interval.set_missed_tick_behavior(MissedTickBehavior::Delay);
+
+        loop {
+            tokio::select! {
+
+                biased;
+
+                _ = &mut pinned => break,
+
+                _ = interval.tick() => self.storage.save().await?,
+
+                Some(next_block_notification) = self.queue_front.recv() => {
+                    let walker = Walker::new(&mut self.storage);
+                    if let Err(_err) = walker.walk(next_block_notification.clone()).await {
+                        // return the block to the back of the task queue to process again later
+                        let _ = self.queue_rear.send(next_block_notification);
+                    }
+                },
+
+            }
+        }
+
+        Ok(())
+    }
+}
+
+impl<'a, S> Walker<'a, S>
+where
+    S: Storage + Clone + Send + Sync + 'static,
+{
+    fn new(storage: &'a mut BlockProcessorStorage<S>) -> Self {
+        Self {
+            path: Vec::new(),
+            visited: HashSet::new(),
+            storage,
+        }
+    }
+
+    async fn walk(mut self, block: BlockId) -> Result<(), ExporterError> {
+        if self.is_block_indexed(&block).await? {
+            return Ok(());
+        }
+
+        let node_visitor = self.get_processed_block_node(&block).await?;
+        self.path.push(node_visitor);
+        while let Some(mut node_visitor) = self.path.pop() {
+            if self.visited.contains(&node_visitor.node.block) {
+                continue;
+            }
+
+            // resolve ancestors
+            let parent = node_visitor.get_parent();
+            if let Some(parent) = parent {
+                if !self.is_block_indexed(&parent).await? {
+                    let parent_node = self.get_processed_block_node(&parent).await?;
+                    self.path.push(node_visitor);
+                    self.path.push(parent_node);
+                    continue;
+                }
+            }
+
+            // resolve siblings
+            let sibling = node_visitor.node.siblings.get(node_visitor.sibling_branch);
+            if let Some(sibling) = sibling {
+                node_visitor.sibling_branch += 1;
```

Comment:
Adding a `fn next_sibling(&mut self) -> Option<BlockId>` to `NodeVisitor` which increments the `sibling_branch` for `Some(_)` will make it more readable.

Reply:
If you do what I described here (https://github.com/linera-io/linera-protocol/pull/3896/files#r2129097593) then there won't be a need for `fn next_sibling` method either.